modules/apps/foundation/frontend-js/frontend-js-loader-modules-extender/src/main/java/com/liferay/frontend/js/loader/modules/extender/npm/model/JSModuleAdapter.java

```diff
@@ -23,10 +23,24 @@
 import java.util.List;
 
 /**
+ * A basic incomplete implementation of {@link JSModule} that can be reused by
+ * real implementations.
  * @author Iván Zaera
  */
 public abstract class JSModuleAdapter implements JSModule {
 
+	/**
+	 *
+	 * @param jsPackage the package where this module belongs
+	 * @param name the name of the module
+	 * @param url the canonical URL of this module
+	 * @param resolvedURL the URL that gives access to the module to which this
+	 *                    module is resolved (in the end, that can be this same
+	 *                    module or another copy of this module contained in a
+	 *                    different bundle)
```

Comment:
OK. The thing is that you may have several copies of the same JSModule in different bundles of the Portal. For example, let's say that you have a module named `p@1.0.0/m` (the module `m` living in package `p, version 1.0.0`) and it is contained in 3 OSGi bundles: b1.jar, b2.jar and b3.jar.

In this scenario you would have 3 JSModule objects inside the NPMRegistry (one per bundle) and one JSResolvedModule (which depending on the algorithm used, would point to one of the other 3). 

The resolved module is used to serve it to the browser, i.e., it's a virtual entity that is passed to the browser so that it uses it instead of passing the browser any of the other three modules.

The final figure would be something like this:

Modules:
-b1.jar:p@1.0.0/m which can be retrieved through its own URL (kind of .../b1.jar/p@1.0.0/m)
-b2.jar:p@1.0.0/m which can be retrieved through its own URL (kind of .../b2.jar/p@1.0.0/m)
-b3.jar:p@1.0.0/m which can be retrieved through its own URL (kind of .../b3.jar/p@1.0.0/m)

Resolved module:
-p@1.0.0/m which can be retrieved through its own URL (kind of .../p@1.0.0/m)

As you can see, the URL of the resolved module doesn't show any reference to any bundle. Also, when you request the resolved module, it is internally resolved to one of the other three, so the requester would end up getting -let's say- .../b2.jar/p@1.0.0/m but she wouldn't notice anything as it is transparent to the requester.